[mlir][bufferization] Better analysis around allocs and block arguments

matthias-springer opened the pull request:
Values that are the result of buffer allocation ops are guaranteed to *not* be the same allocation as block arguments of containing blocks. This fact can be used to allow for more aggressive simplification of `bufferization.dealloc` ops.

maerhart commented on mlir/lib/Dialect/Bufferization/Transforms/BufferDeallocationSimplification.cpp:
```diff
@@ -49,19 +49,53 @@ static LogicalResult updateDeallocIfChanged(DeallocOp deallocOp,
+/// Given a memref value, return the "base" value by skipping over all
+/// ViewLikeOpInterface ops (if any) in the reverse use-def chain.
+static Value getViewBase(Value value) {
+  while (value) {
+    auto viewLikeOp = value.getDefiningOp<ViewLikeOpInterface>();
+    if (!viewLikeOp)
+      return value;
+    value = viewLikeOp.getViewSource();
+  }
+  return value;
```
`value` shouldn't ever be a null value, right? Couldn't we do the following then:
```suggestion
  while (auto viewLikeOp = value.getDefiningOp<ViewLikeOpInterface>())
    value = viewLikeOp.getViewSource();
  return value;
```

maerhart reviewed (approved):
Nice addition!

matthias-springer committed:
[mlir][bufferization] Better analysis around allocs and block arguments

Values that are the result of buffer allocation ops are guaranteed to *not* be the same allocation as block arguments of containing blocks. This fact can be used to allow for more aggressive simplification of `bufferization.dealloc` ops.
